### ISSUE TITLE
Option to write PhotoPrism metadata to XMP sidecars

### ISSUE THREAD

**Issue by bobobo1618** [photoprism/photoprism#402]
PhotoPrism has features like album creation and image tagging (particularly unique is the ML-based classifier). 

It'd be great if PhotoPrism could be configured to write its metadata to XMP sidecars. In the case of standard metadata like tags and descriptions, this should be recognized by software like Darktable and allow a nice integration between the two different applications.

**Comment by lastzero** [comment]
Started to work on XMP already, see https://github.com/photoprism/photoprism/wiki/XMP

It is in fact extremely extensible / undefined, as the name suggests, so developing a complete and tested solution turned out to be too much work for now. Here's an incomplete list of metadata models supported by / used with XMP (see https://github.com/trimmer-io/go-xmp):

- XMP DublinCore (dc)
- XMP Media Management (xmpMM)
- XMP Dynamic Media (xmpDM)
- XMP Rights (xmpRights)
- XMP Jobs (xmpBJ)
- XMP Paged Text (xmpTPg)
- EXIF v2.3.1 (exif, exifEX)
- Adobe Camera Raw (crs)
- Creative Commons (cc)
- DJI Drones (dji)
- ID3 v2.2, v2.3, v2.4 (id3)
- iXML audio recorder (ixml)
- iTunes/MP4 (itunes)
- ISO/MP4 (mp4)
- Quicktime (qt)
- PhotoMechanic (pm)
- Tiff (tiff)
- Riff (riff)
- Photoshop (ps)
- PDF (pdf)

Maybe easier if you only target specific tools and applications like darktable. If it's only working 95%, you don't want to use it for backups or data exchange with other people (especially in a professional context).

**Comment by graciousgrey** [comment]
Writing metadata back to original files is something we plan to do later, but it needs to be well thought through. See https://github.com/photoprism/photoprism/discussions/1092.

What we already do is creating yml backup files with all the metadata.

**Comment by jmalm** [comment]
I can think of the following complexities:

1. Mapping each Photoprism-internal metadata field to one (or more?) fields when **writing** metadata to file (image or XMP sidecar). (By this I don't mean that _every_ Photoprism-internal metadata field necessarily has to be written each time a field is written to file.) 
2. Mapping metadata to populate specific Photoprism-internal metadata fields when **reading** image or XMP sidecar. (Could be several candidate fields in the image or XMP sidecar for one Photoprism-internal metadata field.)
3. Making sure not to loose or mess up existing metadata in image or XMP sidecar when **updating** image or XMP sidecar
4. Possibly: synchronization issues / conflicts

Are there other things that make this a complex matter?

I don't really see synchronization conflicts as being a Photoprism concern, but I may not have enough respect for the complexity of synchronization. In my mind, whenever a file is **read** by Photoprism, it would overwrite any corresponding internal metadata with metadata found in the file being read, and conversely, whenever Photoprism-internal metadata is **written** to a file, it would replace any mapped field in the destination file. The question of **when** to read and write might be another complexity here, but could perhaps be completely separated?

(By the way, is the Photoprism-internal metadata all found in the `Photo` struct, with sub types (i.e. `Details`, `Camera`, etc.)?)

**Comment by pluton-house** [comment]
Also, when a large-scale change is made (such as renaming a Person) this should update the EXIF data in all the original images (if enabled)

**Comment by antgiant** [comment]
Would it be acceptable to offload the actual reading and writing to the very battle tested and popular [ExifTool](https://exiftool.org/)? 

[ExifTool](https://exiftool.org/)
- [Reads and writes a huge number of metadata formats including XMP](https://exiftool.org/#features). 
- Is [GPL licensed](https://exiftool.org/#license)
- Has a GPL 3 licensed [Go Wrapper](https://github.com/barasher/go-exiftool) already made by the community.

If this is acceptable I think it simplifies this problem to “just” one of mapping out the supported metadata.

**Comment by graciousgrey** [comment]
Yes, we will consider using exiftool when we get to this. See also: 

- https://github.com/photoprism/photoprism/discussions/1092

However, this may not work for all of the file formats we support, so we may also choose xmp.

**Comment by antgiant** [comment]
Hey thanks I had missed that issue. 

I’m confused by your file format comment though. ExifTool reads and writes XMP just fine and the list of [file formats supported by ExifTool](https://exiftool.org/#supported) is enormous. Why not just let ExifTool write the XMP for you, or just use ExifTool’s [default priorities](https://exiftool.org/#writing) which already falls back to XMP when other options (exif & iptc) fail.

**Comment by gotjoshua** [comment]
Is this a decision to implement either exif or xmp exclusively? 

or are we talking about settings and options?

Writing metadata back into the originals is a really high priority for me, and i'm looking into microservices that can do it independently of a photoprism feature.

>  creating yml backup files with all the metadata

one possibility could be to loop through these yml files and sync the exif data based on that.

**Comment by JiCiT** [comment]
Created a script that walks the YAML files and syncs some of the data to to EXIF.  See  https://github.com/photoprism/photoprism/discussions/1092#discussioncomment-10205965.
